File: orig/src/manna.js
import { cacheOrFetch } from './_cache.js';

const fetchUrl = 'https://onemin-prod.herokuapp.com/api/v3/partners/304/restaurants/287/product-categories/with-products?type=web';
const website = 'https://www.mannaetterem.hu/rendeles';
const shortName = 'manna';
const name = 'Manna étterem';
const HETI_MENU = 'Heti menü';

const startedAt = performance.now();
export const fetchManna = async () => {
    try {
        const menu = await cacheOrFetch(shortName, fetchUrl, 'json');

        const hetiMenu = menu.find((m) => m.name === HETI_MENU);

        const result = {
            shortName,
            fetchUrl,
            fetchedIn: 0,
            website,
            name,
            offers: (hetiMenu?.products || []).reduce((acc, product) => {
                let [day, date, ...menu] = product.name.split(' ');
                date = `${new Date().getFullYear()}.${date.substring(1, date.length - 2)}`;
                const offer = acc.find((o) => o.date == date);
                if (!offer) {
                    acc.push({
                        date: date.replaceAll('-', '.'),
                        day,
                        offers: [`${menu.join(' ')}: ${product.description}`],
                    });
                } else {
                    offer.offers.push(`${menu.join(' ')}: ${product.description}`);
                }
                return acc;
            }, []),
        };

        result.fetchedIn = performance.now() - startedAt;

        return result;
    } catch (e) {
        return {
            shortName,
            fetchedIn: performance.now() - startedAt,
            errorReason: e.message,
            name,
            fetchUrl,
            offers: [],
            website,
        };
    }
};

File: src/Restaurants/Manna.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.Json;
using System.Text.Json.Serialization;
using debmenu.Caching;
using debmenu.Providers.Inference;
using Serilog;

namespace debmenu.Restaurants;

[method: SetsRequiredMembers]
public sealed class Manna(
    IInferenceProvider inferenceProvider,
    IHttpClientFactory httpClientFactory,
    ILogger logger,
    IHttpResourceStateStore stateStore,
    IRestaurantResultCache resultCache) : Restaurant(
        "https://onemin-prod.herokuapp.com/api/v3/partners/304/restaurants/287/product-categories/with-products?type=web",
        httpClientFactory,
        inferenceProvider,
        logger,
        [],
        stateStore,
        resultCache)
{
    private const string HETI_MENU = "Heti menü";

    public override async Task<Dictionary<string, List<string>>> GetOffersAsync()
    {
        return await GetOffersWithCachingAsync(() => TextWorkflow());
    }

    protected override async Task<string?> ExtractOffersFromText(string html)
    {
        return await Task.FromResult(html);
    }

    protected override Dictionary<string, List<string>> ParseInferenceResponseAsOffers(string json)
    {
        using var op = CreateTimedOperation(nameof(ParseInferenceResponseAsOffers));
        var menu = JsonSerializer.Deserialize<MannaMenu[]>(json) ?? throw new MannaMenuCannotBeParsedException();

        var hetiMenu = menu.FirstOrDefault(menu => menu.Name == HETI_MENU) ?? throw new MannaHetiMenuNotFoundException();

        Dictionary<string, List<string>> offers = [];

        foreach (var product in hetiMenu.Products)
        {
            if (product.ExpectedAt is null)
            {
                Logger.Warning("Skipping {Name} because there is no ExpectedAt defined", product.Name);
                continue;
            }

            var day = product.ExpectedAt.Value.ToString("yyyy-MM-dd");
            if (!offers.ContainsKey(day))
            {
                offers.Add(day, [product.Description]);
            }
            else
            {
                offers[day].Add(product.Description);
            }
        }

        return offers;
    }

    private sealed class MannaMenu
    {
        [JsonPropertyName("name")]
        public required string Name { get; set; }
        [JsonPropertyName("products")]
        public required MannaProduct[] Products { get; set; }
    }

    private sealed class MannaProduct
    {
        [JsonPropertyName("name")]
        public required string Name { get; set; }
        [JsonPropertyName("description")]
        public required string Description { get; set; }
        [JsonPropertyName("expected_at")]
        public DateTime? ExpectedAt { get; set; }
    }

}
